modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java

```diff
@@ -313,7 +313,7 @@ private int _countMatchedParameters(
 		return matched;
 	}
 
-	private void _ensureOpen() {
+	private synchronized void _ensureOpen() {
```

Comment:
@julschong If one thread triggers `_serviceTracker.open()` here, then all other threads calling any of this class' public API should be blocked by the `synchronized(this)` in the `getSingleton()` method. Why can't the `synchronized` in `getSingleton` protect from this race condition?

Reply:
@dantewang As far as why adding `synchronized` to the method works, I'm not 100% certain. And need your help on this.
My concern is that the lock on the method should be released when the thread exits the method allowing 1 other thread to continue. but in reality, first thread never releases the lock until it is done populating the map

Reply:
@dantewang I looked at it again and debugged further. The issue is that serviceTracker.open() is run twice.
see resubmit: https://github.com/liferay-core-infra/liferay-portal/pull/6928